airbyte-integrations/connectors/source-salesforce/source_salesforce/availability_strategy.py

```diff
@@ -0,0 +1,34 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+import typing
+from typing import Optional, Tuple
+
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http.availability_strategy import HttpAvailabilityStrategy
+from requests import HTTPError, codes
+
+if typing.TYPE_CHECKING:
+    from airbyte_cdk.sources import Source
+
+
+class SalesforceAvailabilityStrategy(HttpAvailabilityStrategy):
+    def handle_http_error(
+        self, stream: Stream, logger: logging.Logger, source: Optional["Source"], error: HTTPError
+    ) -> Tuple[bool, Optional[str]]:
+        """
+        There are several types of Salesforce sobjects that require additional processing:
+          1. Sobjects for which the user, after setting up the data using Airbyte, restricted access,
+             and we will receive 403 HTTP errors.
+          2. There are streams that do not allow you to make a sample using Salesforce `query` or `queryAll`.
+             And since we use a dynamic method of generating streams for Salesforce connector - at the stage of discover,
+             we cannot filter out these streams, so we catch them at the stage of reading data.
+        """
+        error_data = error.response.json()[0]
+        if error.response.status_code in [codes.FORBIDDEN, codes.BAD_REQUEST]:
+            error_code = error_data.get("errorCode", "")
+            if error_code != "REQUEST_LIMIT_EXCEEDED" or error_code == "INVALID_TYPE_FOR_OPERATION":
```

Comment:
I think this just evaluates to `if error_code != "REQUEST_LIMIT_EXCEEDED"`, since `error_code == "INVALID_TYPE_FOR_OPERATION"` is included in that 🤔 

Is what we want here essentially "If Forbidden, or Bad Request with the specific INVALID_TYPE_FOR_OPERATION error, the stream is unavailable"?

This isn't where we want to handle rate limiting, as we could run into that at any point in the stream. We used to do it in `read_records` although I'm not 100% sure it belongs there vs in a customized [should_retry method](https://github.com/airbytehq/airbyte/blob/2e099acc52aeddc172c5ee66ed98426c69449a4a/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L227-L238)

Reply:
Edit, @grubberr pointed out that we handle rate limit exceeded by stopping the sync [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-salesforce/source_salesforce/source.py#L144) :)

Reply:
1. About the `if` condition
   yes, to be honest, this second part of the `if` condition is useless `error_code == "INVALID_TYPE_FOR_OPERATION"`
   I just copy-pasted this `if` block from the `read_records` method and I just decided not to change it.

2. About rate-limiting as I understand we discussed it in slack?

Reply:
> About rate-limiting as I understand we discussed it in slack

yup! looks handled in read_stream 👍🏻

Reply:
> About the if condition

I think it's worth refactoring, but separately is fine in order to avoid introducing unexpected behavior here. But I do think we might be excepting too much with regards to BAD_REQUESTs that aren't due to INVALID_TYPE_FOR_OPERATION.